File: Python - pygame/Laser.py
import math
import Global
from collections import deque
from Entity import *

class Laser(Entity):
  pool = deque()

  @staticmethod
  def spawn(anim, x, y, heading, enemy = False):
    asset = Laser() if len(Laser.pool) == 0 or Laser.pool[0].active else Laser.pool.popleft()
    asset.anim = anim
    asset.x = x
    asset.y = y
    asset.dx = 6 * math.cos(math.radians(-heading))
    asset.dy = 6 * math.sin(math.radians(-heading))
    asset.heading = heading
    asset.active = True
    asset.enemy = enemy

    Laser.pool.append(asset)
    Global.entities.insert(0, asset)

  def update(self):
    super().update()
    if self.x < 0 or self.x > Global.WIDTH or self.y < 0 or self.y > Global.HEIGHT:
      self.active = False # delete the laser pulse when it goes off the screen


  def check_collisions(self):
    if not self.active or self.enemy: return # collisions between enemy laser and ship handled by the ship
    for e in Global.currentEntities:
        self.active = not e.collides_with(self)
        if not self.active: break

  def collides_with(self, other):
    if self.enemy and self.intersects(other):
        self.active = False # enemy laser gets destroyed by collision with ship
        return True
    return False

File: TypeScript - p5/src/Laser.ts
import Astro from "./Astro";
import Animation from "./Animation";
import Entity from "./Entity";

export default class Laser extends Entity {
  enemy: boolean = false;

  static spawn(anim: Animation, x: number, y: number, heading: number, enemy: boolean = false) {
    const pool: Laser[] = [];

    const asset = pool.length === 0 || pool[0].active ? new Laser() : pool.shift()!;
    asset.anim = anim;
    asset.x = x;
    asset.y = y;
    asset.heading = heading;
    asset.dx = 6 * Math.cos(heading);
    asset.dy = 6 * Math.sin(heading);
    asset.enemy = enemy;
    asset.active = true;

    pool.push(asset);
    // make lasers first in line to be drawn so they are underneath everything
    Astro.entities.unshift(asset);
  }

  update() {
    super.update();
    // delete the laser pulse when it goes off the screen
    if (this.x < 0 || this.x > Astro.WIDTH || this.y < 0 || this.y > Astro.HEIGHT) this.active = false;
  }

  checkCollisions() {
    if (this.enemy) return; // collisions between enemy laser and ship are handled by the ship

    for (let e of Astro.currentEntities) {
      this.active = !e.collision(this); // the laser pulse stays active unless it hits something
      if (!this.active) break; // once the laser pulse hits something, stop checking for collisions
    }
  }

  collision(other: Entity) {
    if (this.enemy && super.hits(other)) {
      this.active = false; // enemy laser gets destroyed by collision with ship
      return true;
    }
    return false;
  }
}
